File: DailyCore/Sources/DailyCore/Services/FeedParser.swift
import Foundation

/// Robust multi-flavor XML parser supporting RSS 2.0, Atom, and RDF feeds.
public final class FeedParser: NSObject, XMLParserDelegate, @unchecked Sendable {
    
    private var articles: [NewsArticle] = []
    private var currentElement = ""
    private var currentTitle = ""
    private var currentLink = ""
    private var currentDescription = ""
    private var currentContentEncoded = ""
    private var currentAuthor = ""
    private var currentPubDateStr = ""
    private var currentImageUrl: String?
    private var channelImageUrl: String?
    
    private var isInsideItem = false
    private var isInsideChannelImage = false
    
    private let feed: FeedSource
    
    public init(feed: FeedSource) {
        self.feed = feed
        super.init()
    }
    
    /// Parses raw XML data into an array of NewsArticle items.
    public func parse(xmlData: Data) -> [NewsArticle] {
        articles = []
        
        // Sanitize XML string (e.g. naked ampersands)
        if let xmlString = String(data: xmlData, encoding: .utf8) ?? String(data: xmlData, encoding: .isoLatin1) {
            let sanitized = sanitizeXml(xmlString)
            if let sanitizedData = sanitized.data(using: .utf8) {
                let parser = XMLParser(data: sanitizedData)
                parser.delegate = self
                parser.shouldProcessNamespaces = true
                parser.shouldReportNamespacePrefixes = true
                parser.parse()
            }
        }
        
        // Fallback: If XMLParser returned 0 items due to malformed XML, use regex parser
        if articles.isEmpty, let xmlString = String(data: xmlData, encoding: .utf8) ?? String(data: xmlData, encoding: .isoLatin1) {
            articles = parseUsingRegex(xmlString: xmlString)
        }
        
        return articles
    }
    
    // MARK: - XMLParserDelegate
    
    public func parser(
        _ parser: XMLParser,
        didStartElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?,
        attributes attributeDict: [String: String] = [:]
    ) {
        let name = elementName.lowercased()
        currentElement = name
        
        if name == "item" || name == "entry" {
            isInsideItem = true
            currentTitle = ""
            currentLink = ""
            currentDescription = ""
            currentContentEncoded = ""
            currentAuthor = ""
            currentPubDateStr = ""
            currentImageUrl = nil
            return
        }
        
        if !isInsideItem {
            if name == "image" || name == "logo" || name == "icon" {
                isInsideChannelImage = true
            }
            return
        }
        
        // Atom link element: <link href="..." rel="alternate"/>
        if name == "link" {
            if let href = attributeDict["href"] {
                let rel = attributeDict["rel"]?.lowercased()
                if rel == "enclosure" && attributeDict["type"]?.hasPrefix("image/") == true {
                    currentImageUrl = href
                } else if rel == "alternate" || rel == nil || currentLink.isEmpty {
                    currentLink = href
                }
            }
        }
        
        // Media content or thumbnail: <media:content url="...">
        if name == "content" || name == "thumbnail" {
            if let url = attributeDict["url"], isLikelyImage(url: url) {
                if currentImageUrl == nil {
                    currentImageUrl = url
                }
            }
        }
        
        // RSS enclosure: <enclosure url="..." type="image/..."/>
        if name == "enclosure" {
            if let url = attributeDict["url"] {
                let type = attributeDict["type"]?.lowercased() ?? ""
                if type.isEmpty || type.hasPrefix("image/") || isLikelyImage(url: url) {
                    if currentImageUrl == nil {
                        currentImageUrl = url
                    }
                }
            }
        }
    }
    
    public func parser(_ parser: XMLParser, foundCharacters string: String) {
        if isInsideChannelImage && channelImageUrl == nil && currentElement == "url" {
            channelImageUrl = (channelImageUrl ?? "") + string
            return
        }
        
        guard isInsideItem else { return }
        
        switch currentElement {
        case "title":
            currentTitle += string
        case "link":
            if currentLink.isEmpty {
                currentLink += string.trimmingCharacters(in: .whitespacesAndNewlines)
            }
        case "description", "summary":
            currentDescription += string
        case "encoded", "content":
            currentContentEncoded += string
        case "pubdate", "published", "updated", "date":
            currentPubDateStr += string
        case "creator", "author", "name":
            currentAuthor += string
        default:
            break
        }
    }
    
    public func parser(_ parser: XMLParser, foundCDATA CDATABlock: Data) {
        guard let str = String(data: CDATABlock, encoding: .utf8) else { return }
        guard isInsideItem else { return }
        
        switch currentElement {
        case "title":
            currentTitle += str
        case "description", "summary":
            currentDescription += str
        case "encoded", "content":
            currentContentEncoded += str
        case "creator", "author", "name":
            currentAuthor += str
        default:
            break
        }
    }
    
    public func parser(
        _ parser: XMLParser,
        didEndElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?
    ) {
        let name = elementName.lowercased()
        
        if name == "image" || name == "logo" || name == "icon" {
            isInsideChannelImage = false
            return
        }
        
        if name == "item" || name == "entry" {
            isInsideItem = false
            
            let title = decodeHtmlEntities(currentTitle.trimmingCharacters(in: .whitespacesAndNewlines))
            let link = currentLink.trimmingCharacters(in: .whitespacesAndNewlines)
            let date = parseDate(currentPubDateStr.trimmingCharacters(in: .whitespacesAndNewlines))
            
            // Extract image from description or content if not found in tags
            if currentImageUrl == nil {
                currentImageUrl = extractFirstImage(from: currentContentEncoded) ?? extractFirstImage(from: currentDescription)
            }
            
            let finalImageUrl = optimizeMediumImageUrl(currentImageUrl ?? channelImageUrl ?? feed.iconUrl)
            
            // Clean description snippet
            var cleanDesc = stripHtmlTags(currentDescription)
            if cleanDesc.isEmpty && !currentContentEncoded.isEmpty {
                let contentClean = stripHtmlTags(currentContentEncoded)
                cleanDesc = contentClean.count > 280 ? String(contentClean.prefix(280)) + "..." : contentClean
            }
            
            // Format Author & Publication
            var authorName = currentAuthor.trimmingCharacters(in: .whitespacesAndNewlines)
            var pubName = feed.name
            if let inRange = authorName.range(of: " in ") {
                pubName = String(authorName[inRange.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
                authorName = String(authorName[..<inRange.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
            }
            
            let article = NewsArticle(
                id: link.isEmpty ? UUID().uuidString : link,
                title: title.isEmpty ? "No Title" : title,
                link: link,
                publishDate: date,
                imageUrl: finalImageUrl,
                description: cleanDesc,
                content: currentContentEncoded.isEmpty ? currentDescription : currentContentEncoded,
                author: authorName.isEmpty ? nil : authorName,
                publicationName: pubName,
                publicationIconUrl: feed.iconUrl,
                category: feed.category
            )
            
            articles.append(article)
        }
    }
    
    // MARK: - Regex Fallback Parser
    
    private func parseUsingRegex(xmlString: String) -> [NewsArticle] {
        var results: [NewsArticle] = []
        let itemPattern = "<(?:item|entry)[^>]*>([\\s\\S]*?)</(?:item|entry)>"
        guard let itemRegex = try? NSRegularExpression(pattern: itemPattern, options: .caseInsensitive) else { return results }
        
        let nsString = xmlString as NSString
        let matches = itemRegex.matches(in: xmlString, range: NSRange(location: 0, length: nsString.length))
        
        for match in matches {
            let itemContent = nsString.substring(with: match.range(at: 1))
            
            let title = extractTagValue(tag: "title", from: itemContent)
            var link = extractTagValue(tag: "link", from: itemContent)
            if link.isEmpty {
                link = extractAttribute(attribute: "href", tag: "link", from: itemContent)
            }
            let desc = extractTagValue(tag: "description", from: itemContent)
            let content = extractTagValue(tag: "content:encoded", from: itemContent).isEmpty
                ? extractTagValue(tag: "content", from: itemContent)
                : extractTagValue(tag: "content:encoded", from: itemContent)
            let pubDate = extractTagValue(tag: "pubDate", from: itemContent).isEmpty
                ? extractTagValue(tag: "published", from: itemContent)
                : extractTagValue(tag: "pubDate", from: itemContent)
            let author = extractTagValue(tag: "dc:creator", from: itemContent).isEmpty
                ? extractTagValue(tag: "author", from: itemContent)
                : extractTagValue(tag: "dc:creator", from: itemContent)
            
            var imageUrl = extractAttribute(attribute: "url", tag: "media:content", from: itemContent)
            if imageUrl.isEmpty {
                imageUrl = extractAttribute(attribute: "url", tag: "enclosure", from: itemContent)
            }
            if imageUrl.isEmpty {
                imageUrl = extractFirstImage(from: content) ?? extractFirstImage(from: desc) ?? ""
            }
            
            var cleanDesc = stripHtmlTags(desc)
            if cleanDesc.isEmpty && !content.isEmpty {
                let contentClean = stripHtmlTags(content)
                cleanDesc = contentClean.count > 280 ? String(contentClean.prefix(280)) + "..." : contentClean
            }
            
            results.append(NewsArticle(
                id: link.isEmpty ? UUID().uuidString : link,
                title: decodeHtmlEntities(title),
                link: link,
                publishDate: parseDate(pubDate),
                imageUrl: optimizeMediumImageUrl(imageUrl.isEmpty ? feed.iconUrl : imageUrl),
                description: cleanDesc,
                content: content.isEmpty ? desc : content,
                author: author.isEmpty ? nil : author,
                publicationName: feed.name,
                publicationIconUrl: feed.iconUrl,
                category: feed.category
            ))
        }
        
        return results
    }
    
    // MARK: - Helpers
    
    private func sanitizeXml(_ xml: String) -> String {
        // Replace & with &amp; only if not followed by valid entity
        let pattern = "&(?!(?:amp|lt|gt|quot|apos|#\\d+|#[xX][a-fA-F0-9]+);)"
        return xml.replacingOccurrences(of: pattern, with: "&amp;", options: .regularExpression)
    }
    
    private func isLikelyImage(url: String) -> Bool {
        let lower = url.lowercased()
        return lower.contains(".jpg") || lower.contains(".jpeg") || lower.contains(".png") ||
               lower.contains(".webp") || lower.contains(".gif") || lower.contains("image") || lower.contains("miro.medium.com")
    }
    
    private func extractFirstImage(from html: String) -> String? {
        let pattern = #"<img[^>]+src\s*=\s*["']([^"']+)["']"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { return nil }
        let ns = html as NSString
        guard let match = regex.firstMatch(in: html, range: NSRange(location: 0, length: ns.length)) else { return nil }
        return ns.substring(with: match.range(at: 1))
    }
    
    private func extractTagValue(tag: String, from xml: String) -> String {
        let pattern = "<\(tag)[^>]*>(?:<!\\[CDATA\\[([\\s\\S]*?)\\]\\]>|([\\s\\S]*?))</\(tag)>"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { return "" }
        let ns = xml as NSString
        guard let match = regex.firstMatch(in: xml, range: NSRange(location: 0, length: ns.length)) else { return "" }
        if match.range(at: 1).location != NSNotFound {
            return ns.substring(with: match.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if match.range(at: 2).location != NSNotFound {
            return ns.substring(with: match.range(at: 2)).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return ""
    }
    
    private func extractAttribute(attribute: String, tag: String, from xml: String) -> String {
        let pattern = "<\(tag)[^>]*\(attribute)\\s*=\\s*[\"']([^\"']+)[\"'][^>]*>"
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { return "" }
        let ns = xml as NSString
        guard let match = regex.firstMatch(in: xml, range: NSRange(location: 0, length: ns.length)) else { return "" }
        return ns.substring(with: match.range(at: 1))
    }
    
    private func parseDate(_ str: String) -> Date {
        guard !str.isEmpty else { return Date() }
        
        let formats = [
            "EEE, dd MMM yyyy HH:mm:ss Z",
            "EEE, dd MMM yyyy HH:mm:ss zzz",
            "yyyy-MM-dd'T'HH:mm:ssZ",
            "yyyy-MM-dd'T'HH:mm:ss.SSSZ",
            "yyyy-MM-dd'T'HH:mm:ssXXXXX",
            "yyyy-MM-dd HH:mm:ss",
            "EEE, dd MMM yyyy HH:mm zzz"
        ]
        
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        for f in formats {
            formatter.dateFormat = f
            if let date = formatter.date(from: str) {
                return date
            }
        }
        
        if let iso = ISO8601DateFormatter().date(from: str) {
            return iso
        }
        
        return Date()
    }
    
    private func stripHtmlTags(_ str: String) -> String {
        var clean = str
        clean = clean.replacingOccurrences(of: "</p>", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<br>", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<br/>", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<br />", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
        clean = decodeHtmlEntities(clean)
        clean = clean.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        return clean.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    private func decodeHtmlEntities(_ str: String) -> String {
        var decoded = str
        let entities = [
            ("&quot;", "\""),
            ("&apos;", "'"),
            ("&amp;", "&"),
            ("&lt;", "<"),
            ("&gt;", ">"),
            ("&nbsp;", " ")
        ]
        for (ent, val) in entities {
            decoded = decoded.replacingOccurrences(of: ent, with: val)
        }
        
        // Handle decimal numeric entities: &#539;
        if let regex = try? NSRegularExpression(pattern: "&#(\\d+);") {
            let ns = decoded as NSString
            let matches = regex.matches(in: decoded, range: NSRange(location: 0, length: ns.length)).reversed()
            var mod = decoded
            for m in matches {
                let numStr = ns.substring(with: m.range(at: 1))
                if let code = UInt32(numStr), let scalar = UnicodeScalar(code) {
                    if let targetRange = Range(m.range, in: mod) {
                        mod.replaceSubrange(targetRange, with: String(scalar))
                    }
                }
            }
            decoded = mod
        }
        
        // Handle hex numeric entities: &#x21b;
        if let regex = try? NSRegularExpression(pattern: "&#[xX]([0-9a-fA-F]+);") {
            let ns = decoded as NSString
            let matches = regex.matches(in: decoded, range: NSRange(location: 0, length: ns.length)).reversed()
            var mod = decoded
            for m in matches {
                let hexStr = ns.substring(with: m.range(at: 1))
                if let code = UInt32(hexStr, radix: 16), let scalar = UnicodeScalar(code) {
                    if let targetRange = Range(m.range, in: mod) {
                        mod.replaceSubrange(targetRange, with: String(scalar))
                    }
                }
            }
            decoded = mod
        }
        
        return decoded
    }
    
    private func optimizeMediumImageUrl(_ url: String) -> String {
        guard url.contains("miro.medium.com") else { return url }
        let pattern = "v2/resize:[^/]+(/format:[^/]+)?"
        return url.replacingOccurrences(of: pattern, with: "v2/resize:fit:800", options: .regularExpression)
    }
}

File: Android/core-model/src/main/java/com/intellidream/daily/model/FeedParser.kt
package com.intellidream.daily.model

import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.StringReader
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone
import java.util.UUID
import java.util.regex.Pattern

class FeedParser(private val feed: FeedSource) {
    private var lastParsedArticleDate: Long? = null

    fun parse(xmlString: String): List<NewsArticle> {
        val articles = mutableListOf<NewsArticle>()
        lastParsedArticleDate = null
        val sanitized = sanitizeXml(xmlString)

        try {
            val factory = XmlPullParserFactory.newInstance().apply {
                isNamespaceAware = true
            }
            val parser = factory.newPullParser()
            parser.setInput(StringReader(sanitized))

            var eventType = parser.eventType
            var isInsideItem = false
            var isInsideChannelImage = false

            var currentTitle = ""
            var currentLink = ""
            var currentDescription = ""
            var currentContentEncoded = ""
            var currentAuthor = ""
            var currentPubDateStr = ""
            var currentPublishedStr = ""
            var currentDcDateStr = ""
            var currentUpdatedStr = ""
            var currentImageUrl: String? = null
            var channelImageUrl: String? = null

            var currentTag = ""

            while (eventType != XmlPullParser.END_DOCUMENT) {
                when (eventType) {
                    XmlPullParser.START_TAG -> {
                        val name = parser.name.lowercase(Locale.ROOT)
                        currentTag = name

                        if (name == "item" || name == "entry") {
                            isInsideItem = true
                            currentTitle = ""
                            currentLink = ""
                            currentDescription = ""
                            currentContentEncoded = ""
                            currentAuthor = ""
                            currentPubDateStr = ""
                            currentPublishedStr = ""
                            currentDcDateStr = ""
                            currentUpdatedStr = ""
                            currentImageUrl = null
                        } else if (!isInsideItem) {
                            if (name == "image" || name == "logo" || name == "icon") {
                                isInsideChannelImage = true
                            }
                        } else {
                            if (name == "pubdate") {
                                currentPubDateStr = ""
                            } else if (name == "published") {
                                currentPublishedStr = ""
                            } else if (name == "date" || name == "dc:date") {
                                currentDcDateStr = ""
                            } else if (name == "updated" || name == "atom:updated") {
                                currentUpdatedStr = ""
                            }
                            // Atom link: <link href="..." rel="alternate"/>
                            if (name == "link") {
                                val href = parser.getAttributeValue(null, "href")
                                val rel = parser.getAttributeValue(null, "rel")?.lowercase(Locale.ROOT)
                                val type = parser.getAttributeValue(null, "type")?.lowercase(Locale.ROOT)

                                if (rel == "enclosure" && type?.startsWith("image/") == true && href != null) {
                                    currentImageUrl = href
                                } else if (href != null && (rel == "alternate" || rel == null || currentLink.isEmpty())) {
                                    currentLink = href
                                }
                            }

                            // Media content or thumbnail: <media:content url="...">
                            if (name == "content" || name == "thumbnail") {
                                val url = parser.getAttributeValue(null, "url")
                                if (url != null && isLikelyImage(url) && currentImageUrl == null) {
                                    currentImageUrl = url
                                }
                            }

                            // RSS enclosure: <enclosure url="..." type="image/..."/>
                            if (name == "enclosure") {
                                val url = parser.getAttributeValue(null, "url")
                                val type = parser.getAttributeValue(null, "type")?.lowercase(Locale.ROOT) ?: ""
                                if (url != null && (type.isEmpty() || type.startsWith("image/") || isLikelyImage(url))) {
                                    if (currentImageUrl == null) {
                                        currentImageUrl = url
                                    }
                                }
                            }
                        }
                    }

                    XmlPullParser.TEXT -> {
                        val text = parser.text ?: ""
                        if (isInsideChannelImage && channelImageUrl == null && currentTag == "url") {
                            channelImageUrl = (channelImageUrl ?: "") + text
                        } else if (isInsideItem) {
                            when (currentTag) {
                                "title" -> currentTitle += text
                                "link" -> if (currentLink.isEmpty()) currentLink += text.trim()
                                "description", "summary" -> currentDescription += text
                                "encoded", "content" -> currentContentEncoded += text
                                "pubdate" -> currentPubDateStr += text
                                "published" -> currentPublishedStr += text
                                "date" -> currentDcDateStr += text
                                "updated" -> currentUpdatedStr += text
                                "creator", "author", "name" -> currentAuthor += text
                            }
                        }
                    }

                    XmlPullParser.END_TAG -> {
                        val name = parser.name.lowercase(Locale.ROOT)
                        if (name == "image" || name == "logo" || name == "icon") {
                            isInsideChannelImage = false
                        } else if (name == "item" || name == "entry") {
                            isInsideItem = false

                            val title = decodeHtmlEntities(currentTitle.trim())
                            val link = currentLink.trim()
                            
                            // Priority resolution:
                            // 1. pubDate (RSS 2.0 original publication date)
                            // 2. published (Atom 1.0 original publication date)
                            // 3. dc:date (Dublin Core / RDF publication date)
                            // 4. atom:updated / updated (Revision date fallback)
                            val rawPubDate = currentPubDateStr.trim()
                            val rawPublished = currentPublishedStr.trim()
                            val rawDcDate = currentDcDateStr.trim()
                            val rawUpdated = currentUpdatedStr.trim()

                            val parsedDate = parseDateOrNull(rawPubDate)
                                ?: parseDateOrNull(rawPublished)
                                ?: parseDateOrNull(rawDcDate)
                                ?: parseDateOrNull(rawUpdated)

                            val date = if (parsedDate != null) {
                                lastParsedArticleDate = parsedDate
                                parsedDate
                            } else if (lastParsedArticleDate != null) {
                                lastParsedArticleDate!! - 60000L
                            } else {
                                System.currentTimeMillis()
                            }

                            // Extract image from description or content if missing
                            if (currentImageUrl == null) {
                                currentImageUrl = extractFirstImage(currentContentEncoded)
                                    ?: extractFirstImage(currentDescription)
                            }

                            val finalImageUrl = optimizeMediumImageUrl(
                                currentImageUrl ?: channelImageUrl ?: feed.iconUrl
                            )

                            // Clean description snippet
                            var cleanDesc = stripHtmlTags(currentDescription)
                            if (cleanDesc.isEmpty() && currentContentEncoded.isNotEmpty()) {
                                val contentClean = stripHtmlTags(currentContentEncoded)
                                cleanDesc = if (contentClean.length > 280) {
                                    contentClean.take(280) + "..."
                                } else {
                                    contentClean
                                }
                            }

                            // Author in Publication splitting (WinUI / iOS parity)
                            var authorName = currentAuthor.trim()
                            var pubName = feed.name
                            if (authorName.contains(" in ")) {
                                val parts = authorName.split(" in ", limit = 2)
                                authorName = parts[0].trim()
                                pubName = parts[1].trim()
                            }

                            articles.add(
                                NewsArticle(
                                    id = if (link.isNotBlank()) link else UUID.randomUUID().toString(),
                                    title = if (title.isNotBlank()) title else "No Title",
                                    link = link,
                                    publishDate = date,
                                    imageUrl = finalImageUrl,
                                    description = cleanDesc,
                                    content = if (currentContentEncoded.isNotEmpty()) currentContentEncoded else currentDescription,
                                    author = authorName.ifBlank { null },
                                    publicationName = pubName,
                                    publicationIconUrl = feed.iconUrl,
                                    category = feed.category
                                )
                            )
                        }
                        currentTag = ""
                    }
                }
                eventType = parser.next()
            }
        } catch (_: Exception) {
            // Fallback to regex parser on malformed XML
        }

        if (articles.isEmpty()) {
            return parseUsingRegex(xmlString)
        }

        return articles
    }

    private fun parseUsingRegex(xmlString: String): List<NewsArticle> {
        val results = mutableListOf<NewsArticle>()
        val itemPattern = Pattern.compile("<(?:item|entry)[^>]*>([\\s\\S]*?)</(?:item|entry)>", Pattern.CASE_INSENSITIVE)
        val matcher = itemPattern.matcher(xmlString)

        while (matcher.find()) {
            val itemContent = matcher.group(1) ?: continue

            val title = extractTagValue("title", itemContent)
            var link = extractTagValue("link", itemContent)
            if (link.isEmpty()) {
                link = extractAttribute("href", "link", itemContent)
            }
            val desc = extractTagValue("description", itemContent)
            val contentEncoded = extractTagValue("content:encoded", itemContent).ifEmpty {
                extractTagValue("content", itemContent)
            }
            var rawDate = extractTagValue("pubDate", itemContent)
            if (rawDate.isEmpty()) {
                rawDate = extractTagValue("published", itemContent)
            }
            if (rawDate.isEmpty()) {
                rawDate = extractTagValue("dc:date", itemContent)
            }
            if (rawDate.isEmpty()) {
                rawDate = extractTagValue("atom:updated", itemContent)
            }
            if (rawDate.isEmpty()) {
                rawDate = extractTagValue("updated", itemContent)
            }
            val author = extractTagValue("dc:creator", itemContent).ifEmpty {
                val authorBlock = extractTagValue("author", itemContent)
                val innerName = extractTagValue("name", authorBlock)
                if (innerName.isNotEmpty()) innerName else stripHtmlTags(authorBlock)
            }

            var imageUrl = extractAttribute("url", "media:content", itemContent)
            if (imageUrl.isEmpty()) {
                imageUrl = extractAttribute("url", "enclosure", itemContent)
            }
            if (imageUrl.isEmpty()) {
                imageUrl = extractFirstImage(contentEncoded) ?: extractFirstImage(desc) ?: ""
            }

            var cleanDesc = stripHtmlTags(desc)
            if (cleanDesc.isEmpty() && contentEncoded.isNotEmpty()) {
                val contentClean = stripHtmlTags(contentEncoded)
                cleanDesc = if (contentClean.length > 280) contentClean.take(280) + "..." else contentClean
            }

            val parsedDate = parseDateOrNull(rawDate)
            val date = if (parsedDate != null) {
                lastParsedArticleDate = parsedDate
                parsedDate
            } else if (lastParsedArticleDate != null) {
                lastParsedArticleDate!! - 60000L
            } else {
                System.currentTimeMillis()
            }

            results.add(
                NewsArticle(
                    id = if (link.isNotBlank()) link else UUID.randomUUID().toString(),
                    title = decodeHtmlEntities(title),
                    link = link,
                    publishDate = date,
                    imageUrl = optimizeMediumImageUrl(if (imageUrl.isNotBlank()) imageUrl else feed.iconUrl),
                    description = cleanDesc,
                    content = if (contentEncoded.isNotEmpty()) contentEncoded else desc,
                    author = author.ifBlank { null },
                    publicationName = feed.name,
                    publicationIconUrl = feed.iconUrl,
                    category = feed.category
                )
            )
        }

        return results
    }

    private fun sanitizeXml(xml: String): String {
        return xml.replace(Regex("&(?!(?:amp|lt|gt|quot|apos|#\\d+|#[xX][a-fA-F0-9]+);)"), "&amp;")
    }

    private fun isLikelyImage(url: String): Boolean {
        val lower = url.lowercase(Locale.ROOT)
        return lower.contains(".jpg") || lower.contains(".jpeg") || lower.contains(".png") ||
                lower.contains(".webp") || lower.contains(".gif") || lower.contains("image") || lower.contains("miro.medium.com")
    }

    private fun extractFirstImage(html: String): String? {
        val pattern = Pattern.compile("<img[^>]+src\\s*=\\s*[\"']([^\"']+)[\"']", Pattern.CASE_INSENSITIVE)
        val matcher = pattern.matcher(html)
        return if (matcher.find()) matcher.group(1) else null
    }

    private fun extractTagValue(tag: String, xml: String): String {
        val pattern = Pattern.compile("<$tag[^>]*>(?:<!\\[CDATA\\[([\\s\\S]*?)\\]\\]>|([\\s\\S]*?))</$tag>", Pattern.CASE_INSENSITIVE)
        val matcher = pattern.matcher(xml)
        if (matcher.find()) {
            val cdata = matcher.group(1)
            if (cdata != null) return cdata.trim()
            val text = matcher.group(2)
            if (text != null) return text.trim()
        }
        return ""
    }

    private fun extractAttribute(attribute: String, tag: String, xml: String): String {
        val pattern = Pattern.compile("<$tag[^>]*$attribute\\s*=\\s*[\"']([^\"']+)[\"'][^>]*>", Pattern.CASE_INSENSITIVE)
        val matcher = pattern.matcher(xml)
        return if (matcher.find()) matcher.group(1)?.trim() ?: "" else ""
    }

    private fun parseDateOrNull(str: String): Long? {
        if (str.isBlank()) return null
        val cleanStr = str.trim().replace(Regex("\\s+"), " ")

        val formats = listOf(
            "EEE, dd MMM yyyy HH:mm:ss Z",
            "EEE, dd MMM yyyy HH:mm:ss zzz",
            "EEE, dd MMM yyyy HH:mm:ss z",
            "EEE, d MMM yyyy HH:mm:ss Z",
            "EEE, d MMM yyyy HH:mm:ss zzz",
            "d MMM yyyy HH:mm:ss Z",
            "d MMM yyyy HH:mm:ss zzz",
            "yyyy-MM-dd'T'HH:mm:ssZ",
            "yyyy-MM-dd'T'HH:mm:ss.SSSZ",
            "yyyy-MM-dd'T'HH:mm:ssXXX",
            "yyyy-MM-dd'T'HH:mm:ss.SSSXXX",
            "yyyy-MM-dd HH:mm:ss",
            "EEE, dd MMM yyyy HH:mm zzz",
            "EEE, dd MMM yyyy HH:mm Z",
            "yyyy-MM-dd"
        )

        for (fmt in formats) {
            try {
                val sdf = SimpleDateFormat(fmt, Locale.US)
                sdf.timeZone = TimeZone.getTimeZone("UTC")
                val date = sdf.parse(cleanStr)
                if (date != null) return date.time
            } catch (_: Exception) {}
        }

        try {
            return java.time.Instant.parse(cleanStr).toEpochMilli()
        } catch (_: Exception) {}

        return null
    }

    private fun parseDate(str: String): Long {
        return parseDateOrNull(str) ?: System.currentTimeMillis()
    }

    companion object {
        fun stripHtmlTags(str: String): String {
            var clean = str
                .replace(Regex("</p>", RegexOption.IGNORE_CASE), " ")
                .replace(Regex("<br\\s*/?>", RegexOption.IGNORE_CASE), " ")
                .replace(Regex("<[^>]+>"), "")
            clean = decodeHtmlEntities(clean)
            clean = clean.replace(Regex("\\s+"), " ")
            return clean.trim()
        }

        fun decodeHtmlEntities(str: String): String {
            var decoded = str
            val entities = listOf(
                "&quot;" to "\"",
                "&apos;" to "'",
                "&amp;" to "&",
                "&lt;" to "<",
                "&gt;" to ">",
                "&nbsp;" to " ",
                "&#8216;" to "'",
                "&#8217;" to "'",
                "&#8220;" to "\"",
                "&#8221;" to "\"",
                "&#8211;" to "–",
                "&#8212;" to "—",
                "&hellip;" to "…"
            )
            for ((ent, repl) in entities) {
                decoded = decoded.replace(ent, repl)
            }

            // Decimal numeric entities: &#539;
            decoded = Regex("&#(\\d+);").replace(decoded) { match ->
                val code = match.groupValues[1].toIntOrNull()
                if (code != null) {
                    try {
                        String(Character.toChars(code))
                    } catch (_: Exception) {
                        match.value
                    }
                } else match.value
            }

            // Hex numeric entities: &#x21b;
            decoded = Regex("&#[xX]([0-9a-fA-F]+);").replace(decoded) { match ->
                val code = match.groupValues[1].toIntOrNull(16)
                if (code != null) {
                    try {
                        String(Character.toChars(code))
                    } catch (_: Exception) {
                        match.value
                    }
                } else match.value
            }

            return decoded
        }

        fun optimizeMediumImageUrl(url: String): String {
            if (!url.contains("miro.medium.com")) return url
            return url.replace(Regex("v2/resize:[^/]+(/format:[^/]+)?"), "v2/resize:fit:800")
        }
    }
}
